File: MBE.Domain.Elections/PremiumOverride/AgeBandedPremiumOverrideCalculator.cs
using MBE.Domain.Elections.DataAccess;
using MBE.Domain.Elections.Models;

namespace MBE.Domain.Elections.PremiumOverride
{
    public interface IAgeBandedPremiumOverrideCalculator
    {
        decimal GetPremiumOverride(ElectionData electionData);
    }

    public class AgeBandedPremiumOverrideCalculator
    {
        private readonly ITierCostsAgeBandRepository m_tierCostsAgeBandRepository;
        private ElectionData m_electionData;

        public AgeBandedPremiumOverrideCalculator(ITierCostsAgeBandRepository tierCostsAgeBandRepository)
        {
            m_tierCostsAgeBandRepository = tierCostsAgeBandRepository;
        }

        public decimal GetPremiumOverride(ElectionData electionData)
        {
            m_electionData = electionData;
            var tier = m_tierCostsAgeBandRepository.SelectTierCostsAgeBand(electionData.TierID);
            return tier.FlatRate ? tier.PremiumOverride : CalcPremiumOverrideForNotFlatRateTier(tier);
        }

        private decimal CalcPremiumOverrideForNotFlatRateTier(TierCostsAgeBand tier)
        {
            return tier.Per > 0 ? CalcPremiumOverrideForNotAgeBandedAndNotFlatRateTier(tier) : 0;
        }

        private decimal CalcPremiumOverrideForNotAgeBandedAndNotFlatRateTier(TierCostsAgeBand tier)
        {
            return (tier.PremiumOverride * m_electionData.Coverage) / tier.Per;
        }
    }
}

File: MBE.Domain.Elections.Tests/AgeBandedPremiumOverrideCalculator.fs
namespace MBE.Domain.Elections.Tests
open MBE.Domain.Elections
open MBE.Domain.Elections.Models
open System
open Xunit
open FsUnit.Xunit
open MBE.Domain.Elections.DataAccess
open MBE.Domain.Elections.PremiumOverride

module AgeBandedPremiumOverrideCalculator =

    [<Fact>]
    let ``AgeBanded FlatRate Tier Test``() =
        let tier = new TierCostsAgeBand(TierID = 1, BaseTierID = 1, PremiumOverride = decimal 100.00, FlatRate = true )
        let tierCostsAgeBandRepository = { new ITierCostsAgeBandRepository with member this.SelectTierCostsAgeBand(x) = tier }
        let electionData = new ElectionData (ParentUserID = 1500, PlanID = 1, TierID = 1)

        let actual = (new AgeBandedPremiumOverrideCalculator(tierCostsAgeBandRepository)).GetPremiumOverride(electionData)

        Assert.Equal(decimal 100.00, actual)

    [<Fact>]
    let ``AgeBanded Non-FlatRate Tier With Per=0 Test``() =
        let tier = new TierCostsAgeBand(TierID = 1, BaseTierID = 1, PremiumOverride = decimal 100.00, FlatRate = false, Per = decimal 0 )
        let tierCostsAgeBandRepository = { new ITierCostsAgeBandRepository with member this.SelectTierCostsAgeBand(x) = tier }
        let electionData = new ElectionData (ParentUserID = 1500, PlanID = 1, TierID = 1)

        let actual = (new AgeBandedPremiumOverrideCalculator(tierCostsAgeBandRepository)).GetPremiumOverride(electionData)

        Assert.Equal(decimal 0.00, actual)

    [<Fact>]
    let ``PremiumOverrideCalculator AgeBanded Non-FlatRate Tier With Per > 0 Test``() =
        let tier = new TierCostsAgeBand(TierID = 1, BaseTierID = 1, PremiumOverride = decimal 0.11, FlatRate = false, Per = decimal 100.00 )
        let tierCostsAgeBandRepository = { new ITierCostsAgeBandRepository with member this.SelectTierCostsAgeBand(x) = tier }
        let electionData = new ElectionData (ParentUserID = 1500, PlanID = 1, TierID = 1, Coverage = decimal 8333.00)

        let actual = (new AgeBandedPremiumOverrideCalculator(tierCostsAgeBandRepository)).GetPremiumOverride(electionData)

        Assert.Equal(decimal 9.1663, actual)
